proof/refine/ARM/KHeap_R.thy

```diff
@@ -1444,9 +1519,40 @@ lemma setSchedContext_no_stack_update_corres:
       (* sc_replies_relation *)
     apply (clarsimp simp: sc_replies_relation_def sc_replies_of_scs_def map_project_def scs_of_kh_def)
     apply (drule_tac x=p in spec)
-    by (fastforce simp: typ_at'_def ko_wp_at'_def opt_map_def sc_of_def projectKO_opts_defs)
+    by (auto simp: typ_at'_def ko_wp_at'_def opt_map_def sc_of_def projectKO_opts_defs
+            split: if_splits)
 qed
 
+lemma setSchedContext_no_stack_update_corres:
+  "\<lbrakk>sc_relation sc n sc' \<longrightarrow> sc_relation (f sc) n (f' sc');
+     \<forall>sc. sc_replies sc = sc_replies (f sc); \<forall>sc'. objBits sc' = objBits (f' sc');
+     scReply sc' = scReply (f' sc')\<rbrakk>
+    \<Longrightarrow> corres dc
+         (\<lambda>s. kheap s ptr = Some (kernel_object.SchedContext sc n))
+         (ko_at' sc' ptr)
+            (set_object ptr (kernel_object.SchedContext (f sc) n))
+            (setSchedContext ptr (f' sc'))"
+  apply (rule stronger_corres_guard_imp)
+    apply (erule setSchedContext_sc_relation_corres; clarsimp)
+   apply clarsimp
+  apply (clarsimp simp: state_relation_def sc_replies_relation_def)
+  apply (drule_tac x=ptr and y="sc_replies sc" in spec2)
+  apply (clarsimp simp: vs_all_heap_simps obj_at'_real_def ko_wp_at'_def opt_map_def)
+  done
+
+lemma scReply_update_corres:
```

Comment:
This is used only in once place, and, looking at how it is proved, it seems that we could do without this lemma and just use `setSchedContext_sc_relation_corres` directly instead, maybe?